encoding/dictionary.go

```diff
@@ -0,0 +1,96 @@
+package encoding
+
+import (
+	"encoding/binary"
+	"io"
+
+	"github.com/xichen2020/eventdb/generated/proto/encodingpb"
+)
+
+type dictionaryStringIterator struct {
+	mapping   []string
+	meta      encodingpb.StringMeta
+	reader    io.Reader
+	container []byte
+	curr      string
+	err       error
+	closed    bool
+}
+
+func newDictionaryStringIterator(reader io.Reader, meta encodingpb.StringMeta, mapping []string) ForwardStringIterator {
+	return &dictionaryStringIterator{
+		reader:    reader,
+		meta:      meta,
+		mapping:   mapping,
+		container: make([]byte, 8),
+	}
+}
+
+func (d *dictionaryStringIterator) Next() bool {
+	if d.closed {
+		return false
+	}
+
+	_, d.err = d.reader.Read(d.container[:d.meta.BytesPerRecord])
+	if d.err != nil {
+		return false
+	}
+	idx := int(binary.LittleEndian.Uint64(d.container))
+
+	d.curr = d.mapping[idx]
+
+	return true
+}
+
+func (d *dictionaryStringIterator) Current() string {
+	return d.curr
+}
+
+func (d *dictionaryStringIterator) Err() error {
+	return d.err
+}
+
+// Reset of internal data not supported for dictionary iterator.
+func (d *dictionaryStringIterator) Reset(values []string) {}
+
+func (d *dictionaryStringIterator) Close() error {
+	d.closed = true
+	return nil
+}
+
+// Dictionary encoding strategy is to write
+// and then the bytes for the string.
+func (enc *StringEnc) encodeDictionary(writer io.Writer, meta encodingpb.StringMeta, values RewindableStringIterator, mapping map[string]int64, reverseMapping []string) error {
+	// Write out the reverseMapping first.
+	if err := ProtoEncode(&encodingpb.StringMapping{
```

Comment:
Perhaps `StringEnc` should have an internal field with `*encodingpb.StringMapping` type that allows it to reuse across `Encode` calls?